Move c10/util/Exception.h to torch/standalone

desertfire opened the pull request:
Summary: Move a part of c10/util/Exception.h to torch/standalone, so that TORCH_CHECK can be used there.

Test Plan: CI

Differential Revision: D75785559




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Skylion007 reviewed (commented):
This is a pretty common header file for extensions, we may need a shim to link back to the old location for a while.

Skylion007 commented on c10/macros/Macros.h:
```diff
@@ -2,6 +2,8 @@
 #define C10_MACROS_MACROS_H_
 #include <cassert>
 
+#include <torch/standalone/macros/Macros.h>
```
This is bad... we are creating cyclic dependencies between torch and c10. C10 is a tiny utility library and should be kept as such, even if this is a header only dependency, its breaks encapsulation quite badly here. We may have downstream projects wanting to use c10 without torch. Torchstandalone is welcome to bring in a copy of c10

Skylion007 replied:
True, but it does violate encapsulation a bit, which is not great.

janeyx99 reviewed (commented):
Thanks for taking this on! I have a main question about the TORCH_STANDALONE build flag + how the two TORCH_CHECKs are going to coexist in the codebase.

desertfire commented on torch/standalone/util/Exception.h:
```diff
@@ -0,0 +1,78 @@
+#include <sstream>
+#include <string>
+
+#ifdef TORCH_STANDALONE
```
It will be spit out by AOTI's codegen. It's not meant for end user to build a special version of PyTorch.

desertfire committed:
Move c10/util/Exception.h to torch/standalone (#154998)

Summary:

Move a part of c10/util/Exception.h to torch/standalone, so that TORCH_CHECK can be used there.

Test Plan: CI

Differential Revision: D75785559

desertfire commented on torch/standalone/util/Exception.h:
```diff
@@ -0,0 +1,81 @@
+#else // TORCH_STANDALONE
+#include <c10/util/Exception.h>
+
+#define TORCH_STANDALONE_CHECK(cond, ...) TORCH_CHECK(cond, ##__VA_ARGS__)
```
@janeyx99 , @albanD , this is what I end up with for now. Maybe @janeyx99 can try to unify the two versions in future.

janeyx99 replied:
bikeshedding name here, i think i prefer STANDALONE_TORCH_CHECK

desertfire commented on c10/macros/Export.h:
```diff
@@ -1,11 +1,95 @@
 #ifndef C10_MACROS_EXPORT_H_
 #define C10_MACROS_EXPORT_H_
 
+/* Header file to define the common scaffolding for exported symbols.
```
Changes in this file is just to revert my previous commit, because torch/standalone/macros/Export.h uses different macro namings now.
